Performance Profiler: Remove unnecessary useEffect and useStates to address bugs

p-jackson opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes https://github.com/Automattic/dotcom-forge/issues/9244

## Proposed Changes

- Removes unnecessary uses of `useEffect` and `useState`, which appears to fix issues where the ReactQuery cache isn't being used to persist the state of tests when you switch tabs
- Ensures the traffic light loading ticks advance at even 5 second intervals
- Explicitly keeps track of where we're up to in the retesting process since there's no combination of flags that can be used from the `useUrlBasicMetricsQuery` and `useUrlPerformanceInsightsQuery` that can tell us whether this is fetching existing or new test results.


https://github.com/user-attachments/assets/892f276e-bbb8-415c-8cbc-47b64226a690

## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

* Better UX with fixed retest button and up to date reports

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Go to `calypso.localhost:3000/sites/performance/:site`
* Run a test
* Switch to another tab (mimicking users not willing to wait that long and wants to come back later)
* Switch back to the tab, the tab will be at a loading state saying "Getting your report..."
* Check nothing is broken at `calypso.localhost:3000/speed-test-tool?url=<site>`

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

p-jackson committed:
Do not use hash in query key for basic metrics

The hash does not make up part of the request body and so doesn't need
to be in the query key.

p-jackson committed:
Remove unnecessary useEffect and useStates to address bugs

candy02058912 commented:
I haven't looked into the code, but while testing, I noticed that this triggers a new test if the user leaves the tab and comes back, which might set off more tests than required, and may potentially cause load on the backend.


https://github.com/user-attachments/assets/6a0c3b7c-f0db-424d-973e-b6a321cf5b03



If that's something expected and can be fixed in a follow-up, feel free to disregard this comment.

vykes-mac commented:
Is it just me or there seem to be a delay when clicking the `Test again` button? 

https://github.com/user-attachments/assets/ebd18e41-0140-4bb6-9d7e-e86d5863af66

p-jackson commented:
> Is it just me or there seem to be a delay when clicking the Test again button?

@vykes-mac yeah there's definitely a delay. It's the same issue as the loading state not working correctly yet, because that delay is the time it takes for the `/basic` endpoint to queue up a new test.

p-jackson committed:
Explicitly keep track of retest state

p-jackson committed:
Typo

p-jackson committed:
Ensure loading state advances at even intervals of 5 seconds

Using the shared useInterval hook to manage timing since it correctly
handles re-renders without restarting the timer.

p-jackson commented:
>  I noticed that this triggers a new test if the user leaves the tab and comes back, which might set off more tests than required, and may potentially cause load on the backend.\

@candy02058912 I didn't notice this issue during development, but it looks like my latest changes don't seem to have this problem.

paulopmt1 commented:
> I noticed that this triggers a new test if the user leaves the tab and comes back

I had the same issue @p-jackson. It's easy to simulate it when just changing the pages. I didn't notice that on prod. I had previously tested both Home and About page on local and prod. As you can see, when I test the local changes of this pr, it always asks for a new test when switching pages:

https://github.com/user-attachments/assets/9ce3e315-9e27-4c47-9912-6952395c2c22

paulopmt1 reviewed (commented):
> I noticed that this triggers a new test if the user leaves the tab and comes back

I had the same issue @p-jackson. It's easy to simulate it when just changing the pages. I didn't notice that on prod. I had previously tested both Home and About page on local and prod. As you can see, when I test the local changes of this pr, it always asks for a new test when switching pages:

https://github.com/user-attachments/assets/9ce3e315-9e27-4c47-9912-6952395c2c22

vykes-mac committed:
save newly tested reports

vykes-mac committed:
remove memoization of report completed flag

candy02058912 reviewed (approved):
Tested this with logged in and logged out performance profiler and it's working.


https://github.com/user-attachments/assets/0db9ad7f-6048-4919-9abd-6b9ceb5b7eef

> [!NOTE]  
> We will not be exploring saving which page the user was originally on in this case.
> I'll create a new issue for the "Test again" button disappears after a delay as a maintenance task. (https://github.com/Automattic/dotcom-forge/issues/9377)
